### PR TITLE
Add end-to-end tests

### PR DESCRIPTION
**Pull request by adrianchifor**
Using https://github.com/bats-core/bats-core for the testing framework and https://github.com/localstack/localstack for local S3 emulation.

- Adding `--s3-force-path-style` flag to support S3 path bucket addressing (endpoint.com/bucket/key vs. bucket.endpoint.com/key). Local S3 stack doesn't support host addressing so we need this flag to run the tests
- Using `tests/run_tests.sh` as the setup, runner and teardown
- `tests/s3.bats` has all tests for S3 integration (upload, delete, get, prewarm) and `tests/memory.bats` for all in-memory functions (get, invalidate). Memory tests run while the local S3 stack is down for safer and more realistic results

### CODE REVIEW

File: tests/run_tests.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+echo -e "\nRunning AWS S3 localstack"
+docker run -d --name localstack-s3 -e SERVICES=s3 -p 4566:4566 localstack/localstack:0.12.2
+echo -e "Waiting 30s for AWS S3 localstack to be ready ..."
+sleep 30
+
+echo -e "Creating test S3 bucket"
+aws --endpoint=http://localhost:4566 s3api create-bucket --bucket test
+
+echo -e "\nRunning cachenator cluster"
+$DIR/../bin/cachenator -port 8080 -peers localhost:8080,localhost:8081,localhost:8082 \
```
Comment (uberspot) [inline tests/run_tests.sh:42]: Maybe restructure these `localhost:8080`... combinations to be able to reuse the vars you have in helpers.sh ? So they're only defined in one file.

File: tests/s3.bats [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  run DELETE "$CACHE/delete?bucket=test&key=folder/subfolder/blob"
+  [[ "$status" -eq 0 ]]
+  [[ "$output" == "200" ]]
+
+  run AWS s3 ls s3://test/folder/subfolder/blob
+  [ "$status" -eq 1 ]
+}
+
+@test "deleting blob prefix from test bucket" {
+  run DELETE "$CACHE/delete?bucket=test&prefix=folder/subfolder"
```
Comment (uberspot) [inline tests/s3.bats:44]: Move bucket name to var across all these functions ?

File: tests/run_tests.sh (hunk 1 above)
Comment (adrianchifor) [reply tests/run_tests.sh:42]: `helpers.sh` is only accessible in .bats files though

Comment (adrianchifor) [reply tests/run_tests.sh:42]: actually I'll source it in the runner as well, nothing in there that should break it

File: tests/run_tests.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
+
+# make test
+
+set -u
+set -eE
+
+[[ -z $(which bats) ]] && echo "bats not found, install: https://github.com/bats-core/bats-core#installation" && exit 1
```
Comment (uberspot) [inline tests/run_tests.sh:8]:
```
# Test if a command exists.
try_command() {
    command -v "$1" >/dev/null 2>&1
}
```
This is a preferred way of checking if commands exist. It's built into shells vs an external binary. 
https://stackoverflow.com/questions/37056192/which-vs-command-v-in-bash

File: tests/s3.bats (hunk 2 above)
Comment (adrianchifor) [reply tests/s3.bats]: Done

File: tests/run_tests.sh (hunk 3 above)
Comment (adrianchifor) [reply tests/run_tests.sh]: Done

**Review by uberspot — APPROVED** [review]

**Pull request merged**